currencies/rate_converter_test.go

```diff
@@ -13,354 +13,273 @@ import (
-		&http.Client{},
-		mockedHttpServer.URL,
-		time.Duration(24)*time.Hour,
-	)
+		assert.Equal(t, rates, currencyConverter.Rates(), formatTestErrorMsg(test.description, "Rates() doesn't return expected rates"))
 
-	// Verify:
-	assert.Equal(t, 1, len(calledURLs), "sync URL should have been called %d times but was %d", 1, len(calledURLs))
-	assert.Equal(t, currencyConverter.LastUpdated(), (time.Time{}), "LastUpdated() shouldn't return a time set")
-	assert.Nil(t, currencyConverter.Rates(), "Rates() should return nil")
+		// Check for the next 1 second that no fetch was triggered
+		time.Sleep(500 * time.Millisecond)
+		assert.Equal(t, lastUpdated, currencyConverter.LastUpdated(), formatTestErrorMsg(test.description, "Currency rates were fetched before the requested duration was passed"))
+	}
 }
 
-func TestInit(t *testing.T) {
-
+func TestNewRateConverterWithNotifier(t *testing.T) {
 	// Setup:
-	mockedHttpServer := httptest.NewServer(http.HandlerFunc(
-		func(rw http.ResponseWriter, req *http.Request) {
-			rw.WriteHeader(http.StatusOK)
-			rw.Write([]byte(
-				`{
-					"dataAsOf":"2018-09-12",
-					"conversions":{
-						"USD":{
-							"GBP":0.77208
-						},
-						"GBP":{
-							"USD":1.2952
-						}
-					}
-				}`,
-			))
-		}),
-	)
-
-	// Execute:
-	expectedTicks := 5
-	ticksTimes := []time.Time{}
-	ticks := make(chan int)
-	currencyConverter := currencies.NewRateConverterWithNotifier(
-		&http.Client{},
-		mockedHttpServer.URL,
-		time.Duration(100)*time.Millisecond,
-		ticks,
-	)
+	rates := &currencies.Rates{}
+	err := json.Unmarshal(getMockRates(), rates)
+	assert.NoError(t, err, "JSON unmarshalling of conversions failed")
+
+	mockServerHandler := func(mockResponse []byte, code int, initialFailCount int,
+		failRespCode int, failRespMessage []byte) http.Handler {
+		called := 0
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			if called < initialFailCount {
+				rw.WriteHeader(failRespCode)
+				rw.Write(failRespMessage)
+			}
+			called++
+			rw.WriteHeader(code)
+			rw.Write(mockResponse)
+		})
+	}
 
-	// Verify:
-	expectedIntervalDuration := time.Duration(100) * time.Millisecond
-	errorMargin := 0.1 // 10% error margin
-	expectedRates := &currencies.Rates{
-		DataAsOf: time.Date(2018, time.September, 12, 0, 0, 0, 0, time.UTC),
-		Conversions: map[string]map[string]float64{
-			"USD": {
-				"GBP": 0.77208,
-			},
-			"GBP": {
-				"USD": 1.2952,
-			},
+	testCases := []struct {
+		mockResponse     []byte
+		mockResponseCode int
+		initialFailCount int
+		failRespCode     int
+		failRespMessage  []byte
+		updateInterval   int
+		fetchURL         string
+		expectedTicks    int
+		description      string
+	}{
+		{
+			mockResponse:     getMockRates(),
+			mockResponseCode: 200,
+			updateInterval:   100,
+			expectedTicks:    2,
+			initialFailCount: 2,
+			failRespCode:     400,
+			failRespMessage:  []byte("Bad request"),
+			description:      "Fetching currency rates successfully",
```

Comment:
Looks like this (and the test below) return a 400 but are described as `Fetching currency rates successfully`. Copy/paste error?